File: source/libs/data/ProjectPacker.cpp
#include "ProjectPacker.h"
#include "assets/AssetRegistry.h"
#include "scenes/SceneManager.h"
#include "scenes/SceneAsset.h"
#include <Protocol.h>
#include <utility>
#include <vector>

ProjectPacker::ProjectPacker(AssetRegistry* assetRegistry,
                             SceneManager* sceneManager,
                             std::shared_ptr<ComponentRegistry> componentRegistry)
  : m_assetRegistry(assetRegistry),
    m_sceneManager(sceneManager),
    m_componentRegistry(std::move(componentRegistry))
{}

void ProjectPacker::pack(net::Message& message) const
{
  // Same shape as ProjectSerializer::serialize: the flat file assets, then every scene's object tree,
  // then the current scene uuid.
  m_assetRegistry->pack(message);

  const auto& scenes = m_sceneManager->getScenes();
  message.write(static_cast<uint32_t>(scenes.size()));
  for (const auto& [uuid, scene] : scenes)
  {
    scene->pack(message);
  }

  const auto currentScene = m_sceneManager->getCurrentScene();
  message.writeString(currentScene ? uuids::to_string(currentScene->getUUID()) : "");
}

void ProjectPacker::unpack(const net::Message& message) const
{
  net::MessageReader reader(message);

  // Atomic load (mirrors ProjectSerializer::deserialize): parse everything that can throw into local
  // instances first, while the live AssetRegistry/SceneManager are untouched.
  AssetRegistry parsedAssets;
  parsedAssets.unpack(reader);

  const uint32_t sceneCount = reader.read<uint32_t>();
  std::vector<std::shared_ptr<SceneAsset>> parsedScenes;
  parsedScenes.reserve(sceneCount);
  for (uint32_t i = 0; i < sceneCount; ++i)
  {
    parsedScenes.push_back(SceneAsset::unpack(reader, m_componentRegistry));
  }

  const std::string currentSceneUUID = reader.readString();

  // Commit phase: nothing below throws, so the swap to the new project is all-or-nothing. A fresh
  // snapshot must REPLACE the current state (the registries are keyed and won't overwrite), so clear
  // first.
  m_sceneManager->clear();
  *m_assetRegistry = std::move(parsedAssets);

  for (const auto& scene : parsedScenes)
  {
    m_sceneManager->addScene(scene);

    // Register the scene as an asset too (matches deserialize), so it shows up in the editor's asset
    // browser. AssetRegistry::pack ignores Scene records, so this doesn't double-pack.
    m_assetRegistry->registerAsset({ .uuid = scene->getUUID(), .type = AssetType::Scene, .path = scene->getName() });
  }

  if (!currentSceneUUID.empty())
  {
    if (const auto parsed = uuids::uuid::from_string(currentSceneUUID); parsed.has_value())
    {
      if (const auto scene = m_sceneManager->getScene(parsed.value()))
      {
        m_sceneManager->loadScene(scene);
      }
    }
  }
}

File: source/apps/webClient/src/data/ProjectPacker.ts
// Port of source/libs/data/ProjectPacker.{h,cpp} - unpack half only.
//
// pack() is not ported: this client never sends project state back. The atomic-load property is
// preserved and matters here as much as upstream - everything that can throw is parsed into local
// instances first, so a malformed snapshot leaves the currently-rendered project intact rather than
// blanking the view.

import type { MessageReader } from "../net/Protocol";
import { AssetRegistry, AssetType } from "./assets/AssetRegistry";
import type { ComponentRegistry } from "./ComponentRegistry";
import { SceneAsset } from "./scenes/SceneAsset";
import type { SceneManager } from "./scenes/SceneManager";

export class ProjectPacker {
  constructor(
    private readonly assetRegistry: AssetRegistry,
    private readonly sceneManager: SceneManager,
    private readonly componentRegistry: ComponentRegistry,
  ) {}

  unpack(reader: MessageReader): void {
    // Parse phase - nothing below touches live state.
    const parsedAssets = new AssetRegistry();
    parsedAssets.unpack(reader);

    const sceneCount = reader.readUint32();
    const parsedScenes: SceneAsset[] = [];
    for (let i = 0; i < sceneCount; ++i) {
      parsedScenes.push(SceneAsset.unpack(reader, this.componentRegistry));
    }

    const currentSceneUUID = reader.readString();

    // Commit phase - nothing below throws, so the swap is all-or-nothing. A fresh snapshot must
    // REPLACE the current state (both registries are keyed and won't overwrite), so clear first.
    this.sceneManager.clear();
    this.assetRegistry.adopt(parsedAssets);

    for (const scene of parsedScenes) {
      this.sceneManager.addScene(scene);

      // Scenes are registered as assets too, matching upstream - AssetRegistry.pack skips Scene
      // records, so this doesn't double-count on the next snapshot.
      this.assetRegistry.registerAsset({
        uuid: scene.getUUID(),
        type: AssetType.Scene,
        path: scene.getName(),
        className: "",
        body: "",
        displayName: "",
      });
    }

    if (currentSceneUUID) {
      const scene = this.sceneManager.getScene(currentSceneUUID);
      if (scene) {
        this.sceneManager.loadScene(scene);
      }
    }
  }
}
